vtrace/platforms/linux.py

```diff
@@ -307,18 +307,27 @@ def platformExec(self, cmdline):
         self.execing = True
         cmdlist = e_cli.splitargs(cmdline)
         os.stat(cmdlist[0])
-        pid = os.fork()
-
+        
+        pid = os.fork() 
+        
+        # Child will be blocked after execv because of trace me!
         if pid == 0:
             v_posix.ptrace(v_posix.PT_TRACE_ME, 0, 0, 0)
-            # Make sure our parent gets some cycles
-            time.sleep(0.1)
             os.execv(cmdlist[0], cmdlist)
             sys.exit(-1)
 
-        if v_posix.ptrace(PT_ATTACH, pid, 0, 0) != 0:
-            raise Exception("PT_ATTACH failed! linux platformExec")
-
+        # Parent will wait for the child
+        for i in xrange(0, 33):
```

Comment:
Lol I'm so lame! :) I have completely missed the part where vtrace it self
is waiting on the tracee!
So my problem was that after platform exec vtrace will do self.wait(). And
since there are no other events pending from the child vdb it self is
blocked. This was not happening with the WNOHANG because it simply did not
wait for the child it did attach before wait returns something meaningful.
This is much better version but still am unable to make the TRACEEXEC to
work.
It would be great if it was working.

Patch is uploaded and available.


# Child will be blocked after execv because of trace me!
        if pid == 0:
            v_posix.ptrace(v_posix.PT_TRACE_ME, 0, 0, 0)
            os.execv(cmdlist[0], cmdlist)
            sys.exit(-1)

        # Parent will wait for the child to do TRACE_ME
        rid, status, rusage = os.wait4(pid, os.WUNTRACED )

        # Only then we attach
        res = v_posix.ptrace(PT_ATTACH, pid, 0, 0)
        if res < 0:
            raise Exception("Can't PTRACE_ATTACH to PID: %d ", pid)

        # since we are attached to our child - a python interpreter
        # make it run until exec occurs. Would be nice to make it work!
        #res = v_posix.ptrace(PT_SETOPTIONS, pid, PT_O_TRACEEXEC, 0)
        #if res < 0:
        #    raise Exception("Can't PTRACE_STEP to PID: %d ", pid)

        # since vtrace will invoke wait() on the tracee after this
        # make sure there is at least one event waiting
        # otherwise we are blocked in vtrace waiting()
        res = v_posix.ptrace(v_posix.PT_STEP, pid, 0, 0)
        if res < 0:
            raise Exception("Can't PTRACE_STEP to PID: %d ", pid)


2015-04-27 17:05 GMT+02:00 Sergei Kostov <bat.serjo@gmail.com>:

> I agree. The WNOHANG could still lead to race conditions if left like this.
>
> 2015-04-27 17:00 GMT+02:00 invisig0th <notifications@github.com>:
>
>> In vtrace/platforms/linux.py
>> <https://github.com/vivisect/vivisect/pull/60#discussion_r29153934>:
>>
>> >              os.execv(cmdlist[0], cmdlist)
>> >              sys.exit(-1)
>> >
>> > -        if v_posix.ptrace(PT_ATTACH, pid, 0, 0) != 0:
>> > -            raise Exception("PT_ATTACH failed! linux platformExec")
>> > -
>> > +        # Parent will wait for the child
>> > +        for i in xrange(0, 33):
>>
>> Under the assumption that as long as we block the child will get cpu time
>> eventually and we are guaranteed to receive a signal from the kernel
>> eventually, we should try to removing WNOHANG and no longer looping. The
>> child will generate his signal to the parent (and block ) at TRACEME and we
>> should be set to go? Thoughts?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/vivisect/vivisect/pull/60/files#r29153934>.
>>
>
>